### PR TITLE
Put type-in-type flag in ugraph.

### LINKED ISSUE #13086: Incomplete type-in-type
~~~coq
(* -*- coq-prog-args: ("-type-in-type"); -*- *)

Fail Definition bad1@{|Set < Set} := Prop.

Set Universe Polymorphism.
Axiom ax : Type.
Inductive I@{u} : Prop := foo : ax@{u} -> I.

Fail Definition bad2@{v} (x:I@{v}) : I@{Set} := x.
~~~

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix #13086.

### CODE REVIEW

File: kernel/uGraph.mli [hunk 1]
```diff
@@ -16,6 +16,15 @@ type t
 val set_cumulative_sprop : bool -> t -> t
 (** Makes the system incomplete. *)
 
+val set_type_in_type : bool -> t -> t
+
+(** When [type_in_type], functions adding constraints do not fail and
+   may instead ignore inconsistent constraints.
```
Comment (SkySkimmer) [inline kernel/uGraph.mli:22]: I'm not sure if this is what we want to do, we could instead shortcut and never add constraints in type-in-type mode.

**Comment by SkySkimmer** [comment]
Not sure what's causing the mtac2 failure. @Janno any clues?

**Comment by Janno** [comment]
> Not sure what's causing the mtac2 failure. @Janno any clues?

Pinging @beta-ziliani because he just worked on that file, I think. Beta, have a look at https://gitlab.com/coq/coq/-/jobs/760793324#L3346 (and lines above) please. It seems something in this PR here breaks your recent soundness fix?

**Comment by beta-ziliani** [comment]
our CI was broken after the zarith dependency broke, so indeed a test failed when we started merging by hand the PRs. it should work now. sorry!

**Comment by SkySkimmer** [comment]
Bench started

**Comment by SkySkimmer** [comment]
~~~

┌─────────────────────────────┬─────────────────────────┬─────────────────────────────────────────────┬─────────────────────────────────────────────┬────────────────────────────────┬───────────────────┐
│                             │      user time [s]      │                 CPU cycles                  │              CPU instructions               │     max resident mem [KB]      │    mem faults     │
│                             │                         │                                             │                                             │                                │                   │
│                package_name │     NEW     OLD PDIFF   │               NEW               OLD PDIFF   │               NEW               OLD PDIFF   │        NEW        OLD  PDIFF   │  NEW  OLD PDIFF   │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│      coq-mathcomp-ssreflect │   50.56   51.05 -0.96 % │      138489460523      138898517340 -0.29 % │      177334252477      177115512421 +0.12 % │     521508     523896  -0.46 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│        coq-mathcomp-algebra │  145.70  146.08 -0.26 % │      405601815136      405311530627 +0.07 % │      523959484451      523551986995 +0.08 % │     561300     566300  -0.88 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│               coq-fiat-core │  105.32  105.51 -0.18 % │      301678949500      301889589785 -0.07 % │      401364789491      401198274940 +0.04 % │     491328     490748  +0.12 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                    coq-hott │  324.10  324.51 -0.13 % │      879182695274      878680243697 +0.06 % │     1425730566106     1427405556125 -0.12 % │     635828     733920 -13.37 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                         coq │  530.73  531.36 -0.12 % │     1483487997941     1482582673401 +0.06 % │     2000620880194     1998787129199 +0.09 % │     599440     590588  +1.50 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                   coq-verdi │  121.20  121.27 -0.06 % │      335737174915      335929788365 -0.06 % │      414084169658      414055900931 +0.01 % │     599304     599284  +0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│              coq-coquelicot │   69.68   69.71 -0.04 % │      191349339355      190942285938 +0.21 % │      245691200349      245436990803 +0.10 % │     583076     583408  -0.06 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│      coq-mathcomp-character │  160.78  160.82 -0.02 % │      446621816814      446573042641 +0.01 % │      618360599896      618213416518 +0.02 % │     721876     724124  -0.31 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                 coq-unimath │ 3997.50 3998.44 -0.02 % │    11118745306381    11124761074682 -0.05 % │    21159007104109    21167302822268 -0.04 % │    1005540    1116620  -9.95 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│            coq-fiat-parsers │  650.16  649.23 +0.14 % │     1820478298778     1817576784211 +0.16 % │     2743320595476     2742730658375 +0.02 % │    2997068    2997900  -0.03 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                 coq-coqutil │   61.05   60.96 +0.15 % │      169864395391      169495795599 +0.22 % │      211185336441      211290797172 -0.05 % │     530360     531000  -0.12 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│              coq-verdi-raft │ 1262.30 1260.31 +0.16 % │     3511979896968     3509399912072 +0.07 % │     4899625576297     4897865677701 +0.04 % │     935780     932816  +0.32 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│       coq-mathcomp-fingroup │   48.32   48.24 +0.17 % │      133621418300      133208414786 +0.31 % │      183913012351      183832038339 +0.04 % │     486312     486364  -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                    coq-corn │ 1630.82 1627.16 +0.22 % │     4541078786960     4535181745926 +0.13 % │     6992750993157     6986408643130 +0.09 % │     855356     830408  +3.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│       coq-mathcomp-solvable │  172.46  172.05 +0.24 % │      479994534666      478847579238 +0.24 % │      676640302436      676091474825 +0.08 % │     688072     691800  -0.54 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                coq-rewriter │  610.33  608.46 +0.31 % │     1693488788066     1688766573517 +0.28 % │     2583219565674     2581584135631 +0.06 % │     989112     995692  -0.66 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│               coq-fourcolor │ 2594.41 2586.20 +0.32 % │     7226534368607     7209961133985 +0.23 % │    11848267864884    11846442921008 +0.02 % │     774908     774744  +0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│      coq-mathcomp-odd-order │ 1070.57 1067.15 +0.32 % │     2981444676761     2972917494417 +0.29 % │     4752026983010     4750016035497 +0.04 % │     994668    1032316  -3.65 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│             coq-fiat-crypto │ 5000.56 4984.50 +0.32 % │    13892113915598    13838293748892 +0.39 % │    24010602658270    24000642432244 +0.04 % │    2440116    2367812  +3.05 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                coq-compcert │  665.67  663.29 +0.36 % │     1851731954529     1850038088933 +0.09 % │     2559200285577     2558542267985 +0.03 % │    1130272    1130408  -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                 coq-bignums │   58.05   57.84 +0.36 % │      160221737711      159916142491 +0.19 % │      215532300235      215489597283 +0.02 % │     474752     471420  +0.71 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                coq-bedrock2 │  224.42  223.58 +0.38 % │      620922473639      618713434428 +0.36 % │     1023342358768     1021462478898 +0.18 % │    1086900    1087052  -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│          coq-mathcomp-field │  292.20  291.09 +0.38 % │      813817358698      809503086078 +0.53 % │     1249177915279     1248319550677 +0.07 % │     831320     825920  +0.65 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│            coq-math-classes │  214.84  213.98 +0.40 % │      599630305052      598002160657 +0.27 % │      829512596193      828972494309 +0.07 % │     515424     519388  -0.76 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│ coq-rewriter-perf-SuperFast │ 1067.31 1062.29 +0.47 % │     2969564145298     2955812340474 +0.47 % │     4652249898875     4651641067059 +0.01 % │    1190112    1070700 +11.15 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│             coq-lambda-rust │ 1266.02 1258.92 +0.56 % │     3523793889429     3503598818581 +0.58 % │     4952658798200     4936048024686 +0.34 % │    1169008    1218732  -4.08 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                coq-coqprime │   81.10   80.52 +0.72 % │      222992590564      221640149497 +0.61 % │      312106638757      311729043295 +0.12 % │     768236     768356  -0.02 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│       coq-engine-bench-lite │  384.15  381.39 +0.72 % │     1080374163437     1072821788061 +0.70 % │     1847626623395     1839225204155 +0.46 % │    3507372    3507276  +0.00 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                  coq-geocoq │ 1479.26 1468.12 +0.76 % │     4121029742046     4094627416848 +0.64 % │     6230378784959     6226349941601 +0.06 % │    1122364    1179388  -4.84 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                   coq-color │  538.65  534.11 +0.85 % │     1498193227150     1492569973965 +0.38 % │     1901161358656     1899244091412 +0.10 % │    1421444    1410604  +0.77 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│                   coq-flocq │  160.61  159.10 +0.95 % │      444112273817      441886020959 +0.50 % │      552373235015      552070094116 +0.05 % │     821936     822032  -0.01 % │    0    0  +nan % │
├─────────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼────────────────────────────────┼───────────────────┤
│  coq-performance-tests-lite │  553.64  544.67 +1.65 % │     1547085639690     1522255812751 +1.63 % │     2801345002134     2791984608902 +0.34 % │     792672     747100  +6.10 % │    0    0  +nan % │
└─────────────────────────────┴─────────────────────────┴─────────────────────────────────────────────┴─────────────────────────────────────────────┴────────────────────────────────┴───────────────────┘

INFO: failed to install coq-sf-plf
~~~

**Review by ppedrot — APPROVED** [review]
I am fine with the code change although I am not sure this is really what we want.

File: kernel/uGraph.mli (hunk 1 above)
Comment (ppedrot) [reply kernel/uGraph.mli:22]: FTR, both behaviours make sense. We would need a spec of type-in-type to decide which one is best suited...

**Comment by ppedrot** [comment]
Merging soon.

**Pull request merged**